File: solutions/cpp/acronym/1/acronym.cpp
#include "acronym.h"

namespace acronym {
     auto is_separator(char ch) -> bool {
           return std::isspace(ch) or ch == '-';
     } 
        
     auto acronym(std::string expression) -> std::string {
           expression.erase(
               std::remove_if(
                    expression.begin(), 
                    expression.end(),
                    [](auto ch) {
                        return std::ispunct(ch) and 
                            not (ch == '-');
                    }), 
                    expression.end());
                  
            std::string unformatted_acronym {static_cast<char>(std::toupper(expression.front()))};
            bool separator {};
                  
            for (const auto& ch : expression) {
                if (is_separator(ch)) {
                     separator = true;
                }
                else if (separator and std::isalpha(ch)) {
                     unformatted_acronym.push_back(
                         std::toupper(ch)
                     );
                     separator = false;
                }
            }
                
            return unformatted_acronym;
       }

}

File: solutions/haskell/acronym/1/src/Acronym.hs
module Acronym (abbreviate) where

import Data.Char(toUpper, isAlpha, isSpace, isLower, isUpper)

abbreviate :: String -> String
abbreviate = go True
  where 
    go _ [] = []
    go True (x:xs)
      | isAlpha x = toUpper x : go False xs
      | otherwise = go True xs
    go False (_:[]) = []
    go False (x:y:xs)
      | isSpace x || x == '-' = go True (y:xs)
      | isLower x && isUpper y = y : go False xs
      | otherwise = go False (y:xs)
      
      
      
      
